Extensions/ArtifactEngine/Store/artifactItemStore.ts

```diff
@@ -47,13 +62,22 @@ export class ArtifactItemStore {
         return undefined;
     }
 
-    public updateState(item: models.ArtifactItem, state: models.TicketState) {
+    public updateState(item: models.ArtifactItem, state: models.TicketState, location?: boolean) {
         var processedItem = this._downloadTickets.find(x => x.artifactItem.path === item.path);
         if (processedItem) {
             processedItem.state = state;
+            if(location) {
+                processedItem.downloadedFromCache = location;
+            }
             if (state != models.TicketState.InQueue && state != models.TicketState.Processing) {
                 processedItem.finishTime = new Date();
             }
+            if((item.fileHash && item.fileHash === this.hashMap[item.path.substring(this._artifactName.length+1)]) || !this.hashMap[item.path.substring(this._artifactName.length+1)]) {
```

Comment:
do not use hashMap here that should be only used at add time. Add another property to item downloadedFileHash and make destination provider update that, here we can compare fileHash with the downloadedFileHash. Whole logic should work seemlessly if we integrate an artifact provider which returns item hash as a part of rest call.